src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2221 @@
+  current_time = monotime_absolute_usec();
+
+  /* If we have scheduled padding some time in the future, we want to see what
+     bin we are in at the current time */
+  target_bin_us = (circpad_delay_t)
+                  MIN((current_time - mi->padding_scheduled_at_us),
+                      CIRCPAD_DELAY_INFINITE-1);
+
+  /* We are treating this non-padding cell as a padding cell, so we cancel
+     padding timer, if present. */
+  mi->padding_scheduled_at_us = 0;
+  if (mi->padding_timer_scheduled) {
+    mi->padding_timer_scheduled = 0;
+    timer_disable(mi->padding_timer);
+  }
+
+  /* If we are not in a padding state (like start or end), we're done */
+  if (!state)
+    return CIRCPAD_STATE_UNCHANGED;
+
+  /* If we're enforcing a state length on non-padding packets,
+   * decrement it */
+  if (mi->state_length != CIRCPAD_STATE_LENGTH_INFINITE &&
+      state->length_includes_nonpadding &&
+      mi->state_length > 0) {
+    mi->state_length--;
+  }
+
+  /* Perform the specified token removal strategy */
+  switch (state->token_removal) {
+    case CIRCPAD_TOKEN_REMOVAL_NONE:
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_CLOSEST_USEC:
+      circpad_machine_remove_closest_token(mi, target_bin_us, 1);
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_CLOSEST:
+      circpad_machine_remove_closest_token(mi, target_bin_us, 0);
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_LOWER:
+      circpad_machine_remove_lower_token(mi, target_bin_us);
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_HIGHER:
+      circpad_machine_remove_higher_token(mi, target_bin_us);
+      break;
+    case CIRCPAD_TOKEN_REMOVAL_EXACT:
+      circpad_machine_remove_exact(mi, target_bin_us);
+      break;
+  }
+
+  /* Check our token and state length limits */
+  return circpad_check_token_supply(mi);
+}
+
+/**
+ * Send a relay command with a relay cell payload on a circuit to
+ * the particular hopnum.
+ *
+ * Hopnum starts at 1 (1=guard, 2=middle, 3=exit, etc).
+ *
+ * Payload may be null.
+ *
+ * Returns negative on error, 0 on success.
+ */
+static int
+circpad_send_command_to_hop(origin_circuit_t *circ, int hopnum,
+                            uint8_t relay_command, const uint8_t *payload,
+                            ssize_t payload_len)
+{
+  crypt_path_t *target_hop = circuit_get_cpath_hop(circ, hopnum);
+  int ret;
+
+  /* Check that the cpath has the target hop */
+  if (!target_hop) {
+    log_fn(LOG_WARN,LD_CIRC,
+           "Padding circuit %u has %d hops, not %d",
+           circ->global_identifier,
+           circuit_get_cpath_len(circ), hopnum);
+    return -1;
+  }
+
+  /* Check that the target hop is opened */
+  if (target_hop->state != CPATH_STATE_OPEN) {
+    log_fn(LOG_WARN,LD_CIRC,
+           "Padding circuit %u has %d hops, not %d",
+           circ->global_identifier,
+           circuit_get_cpath_opened_len(circ), hopnum);
+    return -1;
+  }
+
+  log_fn(LOG_INFO,LD_CIRC, "Negotiating padding on circuit %u.",
```

Comment:
That's not always true because we use the same function to send DROP cells, so it's not always negotiating padding.

On this note, please check commit f937559 which improves and adds some logs, which I found useful during testing. As I test this further, I might add/remove more logs.

Reply:
Ok I took this commit but changed the loud LOG_WARNS back to LOG_INFO. 3fbc3749b64516e759ae95c825692c74d2c8e039

One way to do this sort of thing without extra WARNs is make sure all the logs we're interested in have a circpad prefix we can grep for. Another might be to create a different LD instead of LD_CIRC.